kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/resolver/ClusterResolutionResult.java

```diff
@@ -6,109 +6,197 @@
+     * </ol>
+     * @return true if fully resolved
+     */
+    public boolean allReferentsAndClusterFullyResolved() {
+        return allReferentsFullyResolved() && !isClusterResolvedRefsFalse();
     }
 
-    // using wildcards intentionally
+    /**
+     * All referents are fully resolved iff
+     * <ol>
+     *     <li>All references can be retrieved from kubernetes (we have no dangling refs)</li>
+     *     <li>No referent has a ResolvedRefs=False condition, which declares that it has unresolved dependencies</li>
+     * </ol>
+     * @return true if fully resolved
+     */
+    public boolean allReferentsFullyResolved() {
+        return !anyReferentResolvedRefsConditionsFalse() && !anyDanglingReferences();
+    }
+
+    /**
+     * Transforms this result into a ResolvedCluster, it is assumed that the client has checked that this ClusterResolutionResult
+     * has no dangling references, therefore it is safe to transform it into a resolved form.
+     * @return a resolved cluster
+     * @throws IllegalStateException if there are any dangling references
+     */
+    public ResolvedCluster toResolvedCluster() {
+        if (anyDanglingReferences()) {
+            throw new IllegalStateException("cannot construct resolved cluster if any references are dangling");
+        }
+        return new ResolvedCluster(cluster,
+                proxyResolutionResult.referentResource(),
+                filterResolutionResults.stream().map(ResolutionResult::referentResource).toList(),
+                serviceResolutionResult.referentResource(), ingressResolutionResults.stream().map(IngressResolutionResult::toResolvedIngress).toList());
+    }
+
+    /**
+     * Find all dangling references (recursive) from this VirtualKafkaCluster
+     * @return a stream of all dangling references
+     */
+    public Stream<ClusterResolutionResult.DanglingReference> findDanglingReferences() {
+        return allReferentResolutionResults().filter(ResolutionResult::dangling)
+                .map(resolutionResult1 -> new ClusterResolutionResult.DanglingReference(resolutionResult1.referrer(), resolutionResult1.reference()));
+    }
+
+    /**
+     * Find all dangling references (recursive) from the specified referrer to the specified kind
+     * @param fromReferrer a LocalRef identifying the Referrer
+     * @param toKind a Referent kind
+     * @return a stream of all dangling references from the specified LocalRef to the specified kind
+     */
     @SuppressWarnings("java:S1452")
-    public Stream<LocalRef<?>> findDanglingReferences(LocalRef<?> from, String kindTo) {
-        Objects.requireNonNull(from);
-        Objects.requireNonNull(kindTo);
-        return danglingReferences.stream().filter(r -> r.from().equals(from) && kindTo.equals(r.to.getKind())).map(DanglingReference::to);
+    public Stream<LocalRef<?>> findDanglingReferences(LocalRef<?> fromReferrer, String toKind) {
```

Comment:
It's very confusing to overload a method with one that has a different result type.

Reply:
fixed, they all return `Stream<DanglingReference>` now